### ISSUE TITLE
Planned incompatible API changes for Godot 4.0

### ISSUE THREAD

**Issue by 2shady4u** [2shady4u/godot-sqlite#71]
I've been working on the GDExtension port of this plugin on a separate branch which can be found [here](https://github.com/2shady4u/godot-sqlite/tree/gd-extension) 

This issue aims to serve as a compiled list of all planned changes that will (possibly) break backwards compatibility:

- Implementation of #60 as duplicating the query_result yourself is a hassle in most cases and not clear for beginners. However, the extent to which this is to be implemented is still to be straightened out...
- ....

Other issues that are to be fixed and will be back-ported:

- Better verbose control as discussed in #73 with multiple levels of verbosity for debugging and release purposes.
- Support for multi-statements as discussed in #58, which requires the implementation of proper rollback behavior.
- ...

## How to test the latest Godot 4.0 version of this plugin?

1. Checkout the latest "gd-extension"-branch on your device.
2. Get the binaries and place them in the `bin/`-folder of the demo-project. This can be done by either:
  - Going to Github Actions (as found [here](https://github.com/2shady4u/godot-sqlite/actions)) and downloading the latest binaries that corresponds to your desired platform(s).
  - Compiling the plugin yourself by following the instructions as found [here](https://github.com/2shady4u/godot-sqlite#how-to-contribute).
3. Open the demo-project using the latest Godot 4 alpha executable and start the demo-project

## How to port my Godot 3.x project to Godot 4.0?

Differences between the API can be easily analyzed by comparing the demo-project's differences between versions:
Godot 3.x: https://github.com/2shady4u/godot-sqlite/blob/master/demo/database.gd
Godot 4.0: https://github.com/2shady4u/godot-sqlite/blob/gd-extension/demo/database.gd

In summary:
- The SQLite plugin doesn't have to be preloaded any longer. As a result, following lines of code:
```gdscript
const SQLite = preload("res://addons/godot-sqlite/bin/gdsqlite.gdns")
var db

func ready():
    db = SQLite.new()
```
has to be replaced by:
```gdscript
var db

func ready():
    db = SQLite.new()
```

- The `verbose_mode`-property has been removed in favor of the `verbosity_level`-property. As a result, following lines of code:
```gdscript
func ready():
    db = SQLite.new()
    db.verbose_mode = True
```
has to be replaced by:
```gdscript
func ready():
    db = SQLite.new()
    db.verbosity_level = SQLite.VERBOSE
```

**Comment by 2shady4u** [comment]
Hi @Barskt!

The binary files found on that branch are the **old** binary files which do not work with Godot 4.0 :(
I'll try and see if I can set up some Github Actions to auto-build the plugin for the `gd-extension`-branch.

Regarding compiling the plugin yourself; you won't need to compile Godot from scratch, just the correct `godot-cpp`-branch.

**Comment by Barskt** [comment]
Ah, I see :/ 
As I don't have much experience with compiling I'll wait eagerly for the auto-builds ;P 
Thank you so much for your quick reply and the great work you are doing with this plugin! :)

**Comment by 2shady4u** [comment]
> Ah, I see :/ As I don't have much experience with compiling I'll wait eagerly for the auto-builds ;P Thank you so much for your quick reply and the great work you are doing with this plugin! :)

You can now download the Windows and Linux binaries [here](https://github.com/2shady4u/godot-sqlite/actions/runs/2042541648).
Just clone this repository and place them inside of the `bin/`-folder.

**Comment by Barskt** [comment]
Thank you so much! 
Got the demo to work :)

One note to anyone else who might be interested.
You need to open the gdsqlite.gdextension file and change the name from: libgdsqlite.windows.**debug**.64.dll to libgdsqlite.windows.**release**.64.dll under the [libraries] section.

**Comment by zenasican** [comment]
First off, amazing job. I tried to install on Alpha 6 but I get:

 E 0:00:00:0815   _register_extension_class: Method/function failed.
  <C++ Source>   core/extension/native_extension.cpp:134 @ _register_extension_class()

Where did I go wrong?

**Comment by 2shady4u** [comment]
> First off, amazing job. I tried to install on Alpha 6 but I get:
> 
> E 0:00:00:0815 _register_extension_class: Method/function failed. <C++ Source> core/extension/native_extension.cpp:134 @ _register_extension_class()
> 
> Where did I go wrong?

Hi @zenasican !

I've pushed an update to the gdextension that hopefully fixes this issue for alpha 6.
If I may ask, are you encountering this issue on Windows or on Linux?
Feel free to download the newest/latest binary [here](https://github.com/2shady4u/godot-sqlite/actions/runs/2218877233).

Cheers!

**Comment by zenasican** [comment]
Awesome! I'm running Windows 10 at the moment. I'll try the new update asap.

**Comment by KirillSerogodsky** [comment]
Alpha 7:
ERROR: Condition "!method" is true. Returning: 0
   at: get_builtin_method_hash (core/variant/variant_call.cpp:1172)
ERROR: Error getting method , hash mismatch.
   at: gdnative_variant_get_ptr_builtin_method (core/extension/gdnative_interface.cpp:475)
godot-cpp from master:
https://github.com/KirillSerogodsky/godot-sqlite/actions/runs/2241829670

**Comment by 2shady4u** [comment]
> Alpha 7:
> 
> ```
> ERROR: Condition "!method" is true. Returning: 0
>    at: get_builtin_method_hash (core/variant/variant_call.cpp:1172)
> ERROR: Error getting method , hash mismatch.
>    at: gdnative_variant_get_ptr_builtin_method (core/extension/gdnative_interface.cpp:475)
> ```
> 
> godot-cpp from master: https://github.com/KirillSerogodsky/godot-sqlite/actions/runs/2241829670

Hey @KirillSerogodsky !

I've synced the `godot-cpp` submodule to the correct commit for alpha 7!
Everything should hopefully work again 🤞

**Comment by atngames** [comment]
Hi,
I might not have well understood how to make it work.
I am on ubuntu linux, did download the latest file.so and putted it in the bin directory
Changed the filename in "gdsqlite.gdextension" to match
got the 
```Failed loading resource: res://addons/godot-sqlite/gdsqlite.gdextension. Make sure resources have been imported by opening the project in the editor at least once.``` error
Thank you for your time with all this !
atn

**Comment by 2shady4u** [comment]
> Hi, I might not have well understood how to make it work. I am on ubuntu linux, did download the latest file.so and putted it in the bin directory Changed the filename in "gdsqlite.gdextension" to match got the `Failed loading resource: res://addons/godot-sqlite/gdsqlite.gdextension. Make sure resources have been imported by opening the project in the editor at least once.` error Thank you for your time with all this ! atn

Hi @atngames 

I've updated to gd-extension to include the correct binary entries for Linux.
Unfortunately, I haven't been able to test the Linux binaries as my virtual Ubuntu does weird things when Godot 4 is opened (= it silently crashes).

I've added some small section in the OP to better describe how to get the Godot 4 version of the plugin working.

**Comment by adamgiacomelli** [comment]
You are wonderful, hard-working people - I've managed to get it to work without any issues.
The key is to place the `bin` folder inside the `addons/godot-sqlite` folder and to download the latest artifacts from an action that ran on the `gd-extentions` branch. It is obvious, but not as explicit as it would have to be.

Let me know where I can donate to buy you a beer.

**Comment by git2013vb** [comment]
I'm using linux and I have copied the linux filed (got from here https://github.com/2shady4u/godot-sqlite/actions/runs/3608518073) in x11 folder (renamed it as the old file .so) 
There a need to make other changes in order to use it?  
In my code I use `const SQLite = preload("res://addons/godot-sqlite/bin/gdsqlite.gdns")` and I have an error.
Sorry but I'm not familiar yet with the whole :)

Assuming I'm using the 3.4 version of your plugin in my project and I want to port it in godot 4. what are exactly the steps to do it?
Thanks :)

**Comment by 2shady4u** [comment]
Hi @git2013vb 

I've added a brand-new section to the OP.
I hope it is sufficient to port your project to Godot 4.0 😄

**Comment by git2013vb** [comment]
I downloaded the demo and I tried to run it - I use Debian 11 - 
But I have this error:
![image](https://user-images.githubusercontent.com/13235119/207803635-7924d270-820f-4804-aa41-a6c82533c873.png)

So when I try to run the demo he cannot recognise `SQLite`

**Comment by vickylance** [comment]
@2shady4u  Is there a way we can compile this plugin in its current state for Android?

**Comment by 2shady4u** [comment]
> @2shady4u Is there a way we can compile this plugin in its current state for Android? Or some fixing is pending to be done before it can be compiled?

Hi @vickylance ,
At the moment the plugin compiles for Android.
However, there's some additional work that needs to be done in the godot-cpp repository to make GDExtension for Android.

**Comment by skysphr** [comment]
For anyone else getting "No suitable library found for GDExtension: res://addons/godot-sqlite/gdsqlite.gdextension." while attempting to export on Android, the fix is opening the .gdextension file in a text editor and adding the following lines:

below [libraries]:
```
android.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.arm64.so"
android.template_release.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.arm64.so"
android.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.x86_64.so"
android.template_release.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.x86_64.so"
```

below [dependencies]:
```
android.arm64 = {}
android.template_release.arm64 = {}
android.x86_64 = {}
android.template_release.x86_64 = {}
```

**Comment by vtgreybeard** [comment]
> For anyone else getting "No suitable library found for GDExtension: res://addons/godot-sqlite/gdsqlite.gdextension." while attempting to export on Android, the fix is opening the .gdextension file in a text editor and adding the following lines:
> 
> below [libraries]:
> 
> ```
> android.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.arm64.so"
> android.template_release.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.arm64.so"
> android.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.x86_64.so"
> android.template_release.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.x86_64.so"
> ```
> 
> below [dependencies]:
> 
> ```
> android.arm64 = {}
> android.template_release.arm64 = {}
> android.x86_64 = {}
> android.template_release.x86_64 = {}
> ```

Well, that got me 1 step closer, thanks!
Any suggestions on the next error while trying to explore to android?

```
 platform/android/export/export_plugin.cpp:744 - Cannot determine architecture for library "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.arm64.so". One of the supported architectures must be used as a tag: arm32 arm64 x86_32 x86_64
  editor/export/editor_export_platform.h:173 - Export: Could not export project files.
```

**Comment by skysphr** [comment]
Can you paste the contents of your .gdextension file? Also what architectures are you exporting for? Sorry I didn't notice the comment earlier.

**Comment by vtgreybeard** [comment]
> 

Gladly!
I'm exporting for armeabi-v7a and arm64-v8a, and my .gdextension file is as follows:

```
[configuration]

entry_symbol = "sqlite_library_init"

[libraries]

macos = "res://addons/godot-sqlite/bin/libgdsqlite.macos.template_debug.framework"
macos.template_release = "res://addons/godot-sqlite/bin/libgdsqlite.macos.template_release.framework"
windows.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.windows.template_debug.x86_64.dll"
windows.template_release.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.windows.template_release.x86_64.dll"
linux.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.linux.template_debug.x86_64.so"
linux.template_release.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.linux.template_release.x86_64.so"

android.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.arm64.so"
android.template_release.arm64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.arm64.so"
android.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_debug.x86_64.so"
android.template_release.x86_64 = "res://addons/godot-sqlite/bin/libgdsqlite.android.template_release.x86_64.so"

[dependencies]

android.arm64 = {}
android.template_release.arm64 = {}
android.x86_64 = {}
android.template_release.x86_64 = {}

macos = {}
macos.template_release = {}
windows.x86_64 = {}
windows.template_release.x86_64 = {}
linux.x86_64 = {}
linux.template_release.x86_64 = {}
```

**Comment by cyberpuffin-digital** [comment]
@2shady4u tacking on here as it seems to relate, but I can open another issue if you'd prefer.  I haven't been able to find any other reference to issues reporting a lack of Web export for the new GDExtension.

[Readme.md](https://github.com/2shady4u/godot-sqlite/blob/gd-extension/README.md) doesn't list the Web under supported operating systems, but it does mention the web builds in the "How to Export" section.

Is this a planned feature to be implemented, a breaking compatibility change, or a bug?

**Comment by 2shady4u** [comment]
Support for Web builds has now been added in the latest [release](https://github.com/2shady4u/godot-sqlite/releases/tag/v4.3), but it seems like it is still quite unstable (on the side of Godot).

**Comment by 2shady4u** [comment]
Closing this issue as the port to Godot 4.X+ has now matured enough.
There's still some issues but they are mostly on the side of godot-cpp.
